### PR TITLE
#3951 Fix crash on save when using the Translate brick

### LINKED ISSUE #3951: Crash on save when using the Translate brick in a new install of PB
Steps to reproduce:
1. Install PixieBrix
2. Connect an account
3. Open the page editor
4. Add a Context Menu item
5. Add the Google Translate brick
6. Click Save

Expected outcome:

- Extension is saved correctly

Result:

- Extension "saves", but this error message is shown:
image
- This happens every time Save is clicked, until the blueprints page is opened

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Fixes #3951 
- Adds a guard rail to the `locator.locate()` function to make sure it's only called from the background context
- Refreshes all bricks (blocks, extensionPoints, readers, services) on `Panel` mount in the page editor, instead of just bricks

## Discussion

- Does it make sense to refresh all the brick types like this in either/both places? Or should I make the minimum change possible and just have the create function refresh services before the permissions check?
- Should we look at any other places where only one brick type is fetched and convert everything to refresh the whole registry in each place? This seems more robust, especially to changes in the future, but not sure of, e.g. performance, ramifications of a change like this.

## Team Coordination

- [ ] This PR requires security review

## Checklist

- [ ] Add tests
- [x] Designate a primary reviewer - @twschiller

### CODE REVIEW

File: src/pageEditor/hooks/useCreate.ts [hunk 1]
```diff
@@ -133,6 +133,22 @@ function useCreate(): CreateCallback {
       };
 
       try {
+        try {
+          // Make sure the pages have the latest blocks for when we reactivate below
```
Comment (BLoe) [inline src/pageEditor/hooks/useCreate.ts:137]:
```suggestion
          // Make sure the pages have the latest bricks for when we reactivate below
```

File: src/pageEditor/Panel.tsx [hunk 2]
```diff
@@ -45,9 +44,8 @@ registerDefaultWidgets();
 const Panel: React.VoidFunctionComponent = () => {
   const context = useDevConnection();
 
-  useAsyncEffect(async () => {
-    await blockRegistry.fetch();
-  }, []);
+  // Refresh the brick registry on mount
+  useRefresh({ refreshOnMount: true });
```
Comment (twschiller) [inline src/pageEditor/Panel.tsx:48]: 👍

File: src/pageEditor/hooks/useCreate.ts [hunk 3]
```diff
@@ -133,6 +133,22 @@ function useCreate(): CreateCallback {
       };
 
       try {
+        try {
```
Comment (twschiller) [inline src/pageEditor/hooks/useCreate.ts:136]: 👍

File: src/permissions/index.ts [hunk 4]
```diff
@@ -150,7 +150,9 @@ export async function serviceOriginPermissions(
     return { origins: [] };
   }
 
-  const localConfig = await locator.locate(dependency.id, dependency.config);
+  const localConfig = await locateWithRetry(dependency.id, dependency.config, {
```
Comment (twschiller) [inline src/permissions/index.ts:153]: 👍

**Comment by twschiller** [comment]
> Does it make sense to refresh all the brick types like this in either/both places? Or should I make the minimum change possible and just have the create function refresh services before the permissions check?

It makes sense to fetch when opening the Page Editor (at least doesn't hurt anything)

> Should we look at any other places where only one brick type is fetched and convert everything to refresh the whole registry in each place? This seems more robust, especially to changes in the future,

I think it's a good idea to spot check and see if there might be places with similar problems

>  but not sure of, e.g. performance, ramifications of a change like this.

It would depend on the location. 

Eventually we need to make the brick request a single request to our API instead of 4. There's no real reason to have to make 4 requests. Also, we'll eventually need to provide a way to get new content since timestamp _X_.

Aside from being slow for people access to a lot of bricks right now, making the extra calls isn't a problem especially if they're all done in parallel

**Review by twschiller — APPROVED** [review]

File: src/pageEditor/hooks/useCreate.ts (hunk 3 above)
Comment (fregante) [inline src/pageEditor/hooks/useCreate.ts:136]: Placing a `try` block inside another `try` block is only useful if you expect the `catch` block to throw, which in this case can't. 

It should be `try{}catch{}try{}catch{}` instead

**Pull request merged**